### PR TITLE
Option to skip backprop computation of convolution

### PR DESCRIPTION
**Pull request by niboshi**
Improvements in https://github.com/yuyu2172/chainer

### CODE REVIEW

**Comment by yuyu2172** [comment]
Thanks for checking my forked repository. I assume that you found this branch from my repository that observed performance of Chainer on common convolution neural networks. https://github.com/yuyu2172/chainer-performance-improvements

The first two commits have already been merged #2147, #2148.

**Comment by niboshi** [comment]
@yuyu2172 

> Thanks for checking my forked repository. I assume that you found this branch from my repository that observed performance of Chainer on common convolution neural networks. https://github.com/yuyu2172/chainer-performance-improvements

Yes, exactly.

> You have not written a description about the intent of this PR precisely. I am guessing that you want to add an option that avoids backward computation for data path. Is this correct?

> About the actual commits you have PRed, the first two commits have already been merged to the master #2147, #2148.

Yes, it's correct. As for the commits you mentioned, I realized that when I rebased to master. Thank you for pointing out.

> Also, I think we can think better about the design of this functionality. For example, the convolution link can automatically skip backward path for data with volatile=ON. This suggestion may not be good if API changes in V2 update, so please take that into consideration.

Well, I'm gonna discuss with others.
It's much appreciated if you have any further insights. Thank you!

**Comment by okuta** [comment]
Could you split this PR?
At first, we want to use `no_data_grad` mode. This mode is no performance degradation.

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -236,20 +235,12 @@ def backward_gpu(self, inputs, grad_outputs):
...
-            for i in moves.range(n):
-                gcol_mats[i] = cuda.cupy.dot(W_mat.T, gy_mats[i])
+            gW = cuda.cupy.tensordot(
+                gy, self.col, ((0, 2, 3), (0, 4, 5))).astype(W.dtype,
+                                                             copy=False)
+            gcol = cuda.cupy.tensordot(W, gy, (0, 1)).astype(x.dtype,
+                                                             copy=False)
+            gcol = cuda.cupy.rollaxis(gcol, 3)
 
             gx = conv.col2im_gpu(
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Please use `self.no_data_grad`.

File: chainer/functions/connection/deconvolution_2d.py [hunk 2]
```diff
@@ -262,24 +261,16 @@ def backward_gpu(self, inputs, grad_outputs):
...
 
-            W_mat = W.reshape(in_c, c * kh * kw)
-            col_mats = col.reshape(
-                n, c * kh * kw, in_h * in_w)
-            gx_mats = gx.reshape(n, in_c, in_h * in_w)
-            for i in moves.range(n):
-                gx_mats[i] = W_mat.dot(col_mats[i])
+            gW = cuda.cupy.tensordot(
+                x, col, ([0, 2, 3], [0, 4, 5])).astype(W.dtype, copy=False)
+            gx = cuda.cupy.tensordot(
```
Comment (okuta) [inline chainer/functions/connection/deconvolution_2d.py]: Please use `self.no_data_grad`.

File: chainer/functions/connection/dilated_convolution_2d.py [hunk 3]
```diff
@@ -286,21 +284,12 @@ def backward_gpu(self, inputs, grad_outputs):
...
-            for i in moves.range(n):
-                gcol_mats[i] = cuda.cupy.dot(W_mat.T, gy_mats[i])
-
+            gW = cuda.cupy.tensordot(
+                gy, self.col, ((0, 2, 3), (0, 4, 5))).astype(W.dtype,
+                                                             copy=False)
+            gcol = cuda.cupy.tensordot(W, gy, (0, 1)).astype(x.dtype,
+                                                             copy=False)
+            gcol = cuda.cupy.rollaxis(gcol, 3)
             gx = conv.col2im_gpu(gcol, self.sy, self.sx, self.ph, self.pw,
```
Comment (okuta) [inline chainer/functions/connection/dilated_convolution_2d.py]: Please use `self.no_data_grad`.

File: chainer/links/connection/convolution_2d.py [hunk 4]
```diff
@@ -54,14 +54,16 @@ class Convolution2D(link.Link):
 
     def __init__(self, in_channels, out_channels, ksize, stride=1, pad=0,
                  wscale=1, bias=0, nobias=False, use_cudnn=True,
-                 initialW=None, initial_bias=None, deterministic=False):
+                 initialW=None, initial_bias=None, deterministic=False,
+                 no_data_grad=False):
```
Comment (okuta) [inline chainer/links/connection/convolution_2d.py]: How about following code?
```python
def __call__(self, x):
    no_data_grad = isinstance(x, chainer.Variable)
     if self.has_uninitialized_params:
         with cuda.get_device(self._device_id):
             self._initialize_params(x.shape[1])
    return convolution_2d.convolution_2d(
              x, self.W, self.b, self.stride, self.pad, self.use_cudnn,
              deterministic=self.deterministic, no_data_grad=no_data_grad)
```

**Comment by niboshi** [comment]
@okuta I'm sorry I missed your review. I'll do it.

**Comment by niboshi** [comment]
@okuta If `no_data_grad` is specified, what should be returned as 'gx'?
Currently uninitialized buffer (created by `cupy.empty_like()`) is returned.
I think it's not desirable behaviour.

**Comment by okuta** [comment]
Maybe we can return ``None``.

**Comment by niboshi** [comment]
OK, I'll fix that way.

File: chainer/links/connection/convolution_2d.py (hunk 4 above)
Comment (niboshi) [inline chainer/links/connection/convolution_2d.py]: Do you mean `no_data_grad = not isinstance(x, chainer.Variable)` ?

**Comment by niboshi** [comment]
Please check my reply to your review regarding `no_data_grad` in a link.

**Comment by okuta** [comment]
Please resolve conflicts.

**Comment by niboshi** [comment]
>```no_data_grad = not isinstance(x, chainer.Variable)```

Is this assumption safe?

**Comment by okuta** [comment]
Please check `requires_grad`.

File: chainer/functions/connection/convolution_2d.py [hunk 5]
```diff
@@ -308,7 +320,8 @@ def backward_gpu(self, inputs, grad_outputs):
             return gx, gW, gb
 
 
-def convolution_2d(x, W, b=None, stride=1, pad=0, cover_all=False, **kwargs):
+def convolution_2d(x, W, b=None, stride=1, pad=0, cover_all=False,
+                   no_data_grad=False, **kwargs):
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Maybe, this fixing is not needed.

File: chainer/links/connection/convolution_2d.py [hunk 6]
```diff
@@ -101,7 +100,8 @@ class Convolution2D(link.Link):
     """  # NOQA
 
     def __init__(self, in_channels, out_channels, ksize=None, stride=1, pad=0,
-                 nobias=False, initialW=None, initial_bias=None, **kwargs):
+                 nobias=False, initialW=None, initial_bias=None,
+                 no_data_grad=False, **kwargs):
```
Comment (okuta) [inline chainer/links/connection/convolution_2d.py]: Maybe, this class does not require change.

**Comment by niboshi** [comment]
How abount changing `no_data_grad` to `requires_data_grad`?
It's more straight-forward and double negation can be avoided.

**Comment by okuta** [comment]
How about using `requires_x_grad`?

**Comment by niboshi** [comment]
Changed to `requires_x_grad`.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**